<a href="https://colab.research.google.com/github/gleetx/bootcamp/blob/main/Module%200/Session%206/m0s6nb04_nested_data_example_NBA.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Nested Data Structures Example: NBA Player Analysis

_Version 1.3_

This notebook exercises the concepts of nested data structures in Python.

The intent is to demonstrate for the student a methodology for discovering and manipulating these data structures.

The focus of this notebook is solely on understanding the nested data structures, so that they can be manipulated in some manner.

#### The methodology illustrated here might be required in the format of the first few questions on an exam.

### Context:  We will be using some data from the National Basketball Association's (NBA) statistics API for this exercise. On an exam, the context will be provided, but students don't necessarily need to understand the context in detail, to complete the exam exercises.

The data is called the matchup stats for when a defensive player is guarding an offensive player during the games of an NBA season. The statistics show how the offensive player performed, using various offensive metrics, while being guarded by the defensive player. They include such metrics as the number of games against each other, the total number of minutes that the defensive player guarded the offensive player, points scored, assists, field goal percentage, and the like. The statistics are compiled over the multiple games that the two players' teams played each other over that season.

Analysts compare these matchup statistics with how the offensive player performed over the full season, to assess the defensive player's effectiveness. For example, let's say that the offensive player averaged 20 points per 48 minute game for the season. And then let's say that the offensive player scored 10 points in 48 minutes, while matched up against a particular defender. That defender would be considered very good, because he held the offensive player to half of their points per game.

## Objective

The raw data is in a json file, which comes to us as a dictionary with embedded dictonaries and lists.

For our purposes here, we are simply be working to understand the nested data structure of this json input file.

We assume that this would be the first part of a larger data analysis on the exam.

In [1]:
!wget https://raw.githubusercontent.com/gt-cse-6040/bootcamp/main/Module%200/Session%206/nba_json.txt

--2024-02-19 00:19:24--  https://raw.githubusercontent.com/gt-cse-6040/bootcamp/main/Module%200/Session%206/nba_json.txt
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.108.133, 185.199.109.133, 185.199.110.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.108.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 8903 (8.7K) [text/plain]
Saving to: ‘nba_json.txt’

nba_json.txt        100%[===================>]   8.69K  --.-KB/s    in 0s      

2024-02-19 00:19:24 (45.0 MB/s) - ‘nba_json.txt’ saved [8903/8903]



## The data file

#### nba_json.txt
    This is the raw data, scraped from the NBA API. We will put this into a variable, nested, which will be a list of dictionaries.

Don't worry about the packages or syntax here to import from the json file, as you will learn about these later in the course.

In [2]:
# import the required package for bringing in the json file
import ast

# reading the data from the json file
with open('nba_json.txt') as f:
    data = f.readlines()

nested = [] # this is going to be our nested data structure
#put each line(string) into a dictionary
for line in data:
    result = ast.literal_eval(line)
#     print(result)
    nested.append(result)

## We want to understand the elements of the nested data structure, and how to address each of them.

What is the overall data type of the variable nested?

In [3]:
print(type(nested))

<class 'list'>


#### The variable nested is a list, so let's understand what that means.

We know that a list will have one or more elements, and we use square brackets to address each element in the list. The list indices start at 0 and go up to n-1, where n is the number of elements in the list.

The code below loops through this list and prints the data type of each element in the list.

In [4]:
print(nested)

[{'resource': 'leagueseasonmatchups', 'parameters': {'LeagueID': '00', 'Season': '2019-20', 'SeasonType': 'Regular Season', 'PORound': 0, 'PerMode': 'Totals', 'DefTeamID': None, 'OffTeamID': None, 'OffPlayerID': None, 'DefPlayerID': 202329}, 'resultSets': [{'name': 'SeasonMatchups', 'headers': ['SEASON_ID', 'OFF_PLAYER_ID', 'OFF_PLAYER_NAME', 'DEF_PLAYER_ID', 'DEF_PLAYER_NAME', 'GP', 'MATCHUP_MIN', 'PARTIAL_POSS', 'PLAYER_PTS', 'TEAM_PTS', 'MATCHUP_AST', 'MATCHUP_TOV', 'MATCHUP_BLK', 'MATCHUP_FGM', 'MATCHUP_FGA', 'MATCHUP_FG_PCT', 'MATCHUP_FG3M', 'MATCHUP_FG3A', 'MATCHUP_FG3_PCT', 'HELP_BLK', 'HELP_FGM', 'HELP_FGA', 'HELP_FG_PERC', 'MATCHUP_FTM', 'MATCHUP_FTA', 'SFL', 'MATCHUP_TIME_SEC'], 'rowSet': [['22019', 1629060, 'Rui Hachimura', 202329, 'Al-Farouq Aminu', 1, '6:45', 31.9, 4, 47, 1, 0, 0, 2, 5, 0.4, 0, 0, 0.0, 0, 0, 0, 0, 0, 0, 0, 404.7], ['22019', 1626168, 'Trey Lyles', 202329, 'Al-Farouq Aminu', 1, '7:31', 30.5, 3, 22, 0, 1, 0, 1, 3, 0.333, 1, 2, 0.5, 0, 0, 0, 0, 0, 0, 0, 451.1]

In [5]:
for i in nested:  #loop over the elements of the list
    print(type(i))

<class 'dict'>
<class 'dict'>
<class 'dict'>
<class 'dict'>
<class 'dict'>
<class 'dict'>
<class 'dict'>


#### We see that the list has 7 elements, and that each element of the list is a dictionary.

#### So we have a list of dictionaries.

#### Let's look at the dictionary that is the first element in the list. Pay attention to the notation that we are using.

In [6]:
# display the first dictionary in the list
# Using list notation to address this element
print("nested_first_dict=",nested[0])

nested_first_dict= {'resource': 'leagueseasonmatchups', 'parameters': {'LeagueID': '00', 'Season': '2019-20', 'SeasonType': 'Regular Season', 'PORound': 0, 'PerMode': 'Totals', 'DefTeamID': None, 'OffTeamID': None, 'OffPlayerID': None, 'DefPlayerID': 202329}, 'resultSets': [{'name': 'SeasonMatchups', 'headers': ['SEASON_ID', 'OFF_PLAYER_ID', 'OFF_PLAYER_NAME', 'DEF_PLAYER_ID', 'DEF_PLAYER_NAME', 'GP', 'MATCHUP_MIN', 'PARTIAL_POSS', 'PLAYER_PTS', 'TEAM_PTS', 'MATCHUP_AST', 'MATCHUP_TOV', 'MATCHUP_BLK', 'MATCHUP_FGM', 'MATCHUP_FGA', 'MATCHUP_FG_PCT', 'MATCHUP_FG3M', 'MATCHUP_FG3A', 'MATCHUP_FG3_PCT', 'HELP_BLK', 'HELP_FGM', 'HELP_FGA', 'HELP_FG_PERC', 'MATCHUP_FTM', 'MATCHUP_FTA', 'SFL', 'MATCHUP_TIME_SEC'], 'rowSet': [['22019', 1629060, 'Rui Hachimura', 202329, 'Al-Farouq Aminu', 1, '6:45', 31.9, 4, 47, 1, 0, 0, 2, 5, 0.4, 0, 0, 0.0, 0, 0, 0, 0, 0, 0, 0, 404.7], ['22019', 1626168, 'Trey Lyles', 202329, 'Al-Farouq Aminu', 1, '7:31', 30.5, 3, 22, 0, 1, 0, 1, 3, 0.333, 1, 2, 0.5, 0, 0, 0, 

Let's take this out to Python Tutor to visualize.

[Python Tutor](https://pythontutor.com/visualize.html#mode=display)


#### Note that this first dict in 'nested' is for the player Al-Farouq Aminu, with the player_id 202329.

How would we get the second player in 'nested'?

In [7]:
# display the second dictionary in the list
# Using list notation to address this element
print(nested[1])

{'resource': 'leagueseasonmatchups', 'parameters': {'LeagueID': '00', 'Season': '2019-20', 'SeasonType': 'Regular Season', 'PORound': 0, 'PerMode': 'Totals', 'DefTeamID': None, 'OffTeamID': None, 'OffPlayerID': None, 'DefPlayerID': 1628389}, 'resultSets': [{'name': 'SeasonMatchups', 'headers': ['SEASON_ID', 'OFF_PLAYER_ID', 'OFF_PLAYER_NAME', 'DEF_PLAYER_ID', 'DEF_PLAYER_NAME', 'GP', 'MATCHUP_MIN', 'PARTIAL_POSS', 'PLAYER_PTS', 'TEAM_PTS', 'MATCHUP_AST', 'MATCHUP_TOV', 'MATCHUP_BLK', 'MATCHUP_FGM', 'MATCHUP_FGA', 'MATCHUP_FG_PCT', 'MATCHUP_FG3M', 'MATCHUP_FG3A', 'MATCHUP_FG3_PCT', 'HELP_BLK', 'HELP_FGM', 'HELP_FGA', 'HELP_FG_PERC', 'MATCHUP_FTM', 'MATCHUP_FTA', 'SFL', 'MATCHUP_TIME_SEC'], 'rowSet': [['22019', 202696, 'Nikola Vucevic', 1628389, 'Bam Adebayo', 4, '19:44', 97.5, 22, 100, 8, 2, 0, 10, 22, 0.455, 2, 5, 0.4, 0, 0, 0, 0, 0, 0, 0, 1184.1], ['22019', 203954, 'Joel Embiid', 1628389, 'Bam Adebayo', 4, '18:03', 90.8, 31, 79, 2, 1, 0, 11, 23, 0.478, 2, 5, 0.4, 0, 0, 0, 0, 7, 9, 5, 

#### We can see that this is a dictionary (the squiggly brackets at the very beginning and end) and that it has other nested data elements within the dictionary.

Let's look at the dictionary to see if we can visualize its structure.

We will do this in two steps.

First, we want to know the keys of the dictionary. This will tell us the "top level" data elements.

Then, we will output an element of the list (the dictionary), put it into a markdown cell, and see if we can use CRLF's and indentations to visualize the dictionary and its nested elements.

In [8]:
# First, what are the keys of the dictionary?
# Remember that each element of "nested" is a dictionary
# So nested[i] defines each dictionary in the list of dicts
print(nested[0].keys())


dict_keys(['resource', 'parameters', 'resultSets'])


#### So the dictionary has 3 keys.

Now let's output the first of the list elements from nested, a dictionary.


In [9]:
# Again, note the list notation to address the dictionary
print(nested[0])

{'resource': 'leagueseasonmatchups', 'parameters': {'LeagueID': '00', 'Season': '2019-20', 'SeasonType': 'Regular Season', 'PORound': 0, 'PerMode': 'Totals', 'DefTeamID': None, 'OffTeamID': None, 'OffPlayerID': None, 'DefPlayerID': 202329}, 'resultSets': [{'name': 'SeasonMatchups', 'headers': ['SEASON_ID', 'OFF_PLAYER_ID', 'OFF_PLAYER_NAME', 'DEF_PLAYER_ID', 'DEF_PLAYER_NAME', 'GP', 'MATCHUP_MIN', 'PARTIAL_POSS', 'PLAYER_PTS', 'TEAM_PTS', 'MATCHUP_AST', 'MATCHUP_TOV', 'MATCHUP_BLK', 'MATCHUP_FGM', 'MATCHUP_FGA', 'MATCHUP_FG_PCT', 'MATCHUP_FG3M', 'MATCHUP_FG3A', 'MATCHUP_FG3_PCT', 'HELP_BLK', 'HELP_FGM', 'HELP_FGA', 'HELP_FG_PERC', 'MATCHUP_FTM', 'MATCHUP_FTA', 'SFL', 'MATCHUP_TIME_SEC'], 'rowSet': [['22019', 1629060, 'Rui Hachimura', 202329, 'Al-Farouq Aminu', 1, '6:45', 31.9, 4, 47, 1, 0, 0, 2, 5, 0.4, 0, 0, 0.0, 0, 0, 0, 0, 0, 0, 0, 404.7], ['22019', 1626168, 'Trey Lyles', 202329, 'Al-Farouq Aminu', 1, '7:31', 30.5, 3, 22, 0, 1, 0, 1, 3, 0.333, 1, 2, 0.5, 0, 0, 0, 0, 0, 0, 0, 451.1],

In [11]:
# another way of outputting the variable

#nested[0]

#### Note the differences in how Python formats the two outputs, using print() versus just calling the variable to output.

1.  With print(), you just get the output on lines, with the next line just picking up wherever the text is.

2.  With the variable called directly, though, we can see that it formats the output to show the structure of the nested elements, starting a new row for each new element at that level.

3.  One note on displaying variables in these two manners. The second manner only outputs if it is the last line of the cell. If you want to display a variable in the middle of the cell, with code after it, you must use print().

#### Now let's visualize the dictionary.

We know the three top-level keys, so let's set those first, by putting them onto their own lines.Note that we have truncated the 'resultSets' values, for brevity.

##### resource
    This defines the NBA API endpoint that we called to bring back this data. Simple key-value pair.

##### parameters
    Defines the parameters/variables that we passed to the API endpoint, to tell it which data to select and send back to us. We can see that this is a key, with a dictionary as the value (squiggly braces).

##### resultSets
    This is the player matchup data that we will be putting into our Pandas data frame. We can see that this is a key, with a list as the value (square brackets). We can also see that there appear to be dictionaries and lists within this list, and we will examine this later.


{'resource': 'leagueseasonmatchups',

'parameters': {'LeagueID': '00', 'Season': '2019-20', 'SeasonType': 'Regular Season', 'PORound': 0, 'PerMode': 'Totals', 'DefTeamID': None, 'OffTeamID': None, 'OffPlayerID': None, 'DefPlayerID': 203500},

'resultSets': [{'name': 'SeasonMatchups', 'headers': ['SEASON_ID', 'OFF_PLAYER_ID', 'OFF_PLAYER_NAME', 'DEF_PLAYER_ID', 'DEF_PLAYER_NAME', 'GP', 'MATCHUP_MIN', 'PARTIAL_POSS', 'PLAYER_PTS', 'TEAM_PTS', 'MATCHUP_AST', 'MATCHUP_TOV', 'MATCHUP_BLK', 'MATCHUP_FGM', 'MATCHUP_FGA', 'MATCHUP_FG_PCT', 'MATCHUP_FG3M', 'MATCHUP_FG3A', 'MATCHUP_FG3_PCT', 'HELP_BLK', 'HELP_FGM', 'HELP_FGA', 'HELP_FG_PERC', 'MATCHUP_FTM', 'MATCHUP_FTA', 'SFL', 'MATCHUP_TIME_SEC'], 'rowSet': [['22019', 1629060, 'Rui Hachimura', 202329, 'Al-Farouq Aminu', 1, '6:45', 31.9, 4, 47, 1, 0, 0, 2, 5, 0.4, 0, 0, 0.0, 0, 0, 0, 0, 0, 0, 0, 404.7],}]}

#### Now let's loop over the keys of this dictionary, to display the value of each.

This is just a standard Python loop over a dict.


In [12]:
# loop over the dictionary, display the key and value pair
for k,v in nested[0].items():
    print("key:", k)
    print("value:",v,"\n")


key: resource
value: leagueseasonmatchups 

key: parameters
value: {'LeagueID': '00', 'Season': '2019-20', 'SeasonType': 'Regular Season', 'PORound': 0, 'PerMode': 'Totals', 'DefTeamID': None, 'OffTeamID': None, 'OffPlayerID': None, 'DefPlayerID': 202329} 

key: resultSets
value: [{'name': 'SeasonMatchups', 'headers': ['SEASON_ID', 'OFF_PLAYER_ID', 'OFF_PLAYER_NAME', 'DEF_PLAYER_ID', 'DEF_PLAYER_NAME', 'GP', 'MATCHUP_MIN', 'PARTIAL_POSS', 'PLAYER_PTS', 'TEAM_PTS', 'MATCHUP_AST', 'MATCHUP_TOV', 'MATCHUP_BLK', 'MATCHUP_FGM', 'MATCHUP_FGA', 'MATCHUP_FG_PCT', 'MATCHUP_FG3M', 'MATCHUP_FG3A', 'MATCHUP_FG3_PCT', 'HELP_BLK', 'HELP_FGM', 'HELP_FGA', 'HELP_FG_PERC', 'MATCHUP_FTM', 'MATCHUP_FTA', 'SFL', 'MATCHUP_TIME_SEC'], 'rowSet': [['22019', 1629060, 'Rui Hachimura', 202329, 'Al-Farouq Aminu', 1, '6:45', 31.9, 4, 47, 1, 0, 0, 2, 5, 0.4, 0, 0, 0.0, 0, 0, 0, 0, 0, 0, 0, 404.7], ['22019', 1626168, 'Trey Lyles', 202329, 'Al-Farouq Aminu', 1, '7:31', 30.5, 3, 22, 0, 1, 0, 1, 3, 0.333, 1, 2, 0.5, 0,

In [13]:
# Let's output individual keys
# Notice the hierarchy of how we are addressing the dict, which is an element of the list
print(nested[0]['resource'])
print("\n")
print(nested[0]['parameters'])
print("\n")
print(nested[0]['resultSets'])

leagueseasonmatchups


{'LeagueID': '00', 'Season': '2019-20', 'SeasonType': 'Regular Season', 'PORound': 0, 'PerMode': 'Totals', 'DefTeamID': None, 'OffTeamID': None, 'OffPlayerID': None, 'DefPlayerID': 202329}


[{'name': 'SeasonMatchups', 'headers': ['SEASON_ID', 'OFF_PLAYER_ID', 'OFF_PLAYER_NAME', 'DEF_PLAYER_ID', 'DEF_PLAYER_NAME', 'GP', 'MATCHUP_MIN', 'PARTIAL_POSS', 'PLAYER_PTS', 'TEAM_PTS', 'MATCHUP_AST', 'MATCHUP_TOV', 'MATCHUP_BLK', 'MATCHUP_FGM', 'MATCHUP_FGA', 'MATCHUP_FG_PCT', 'MATCHUP_FG3M', 'MATCHUP_FG3A', 'MATCHUP_FG3_PCT', 'HELP_BLK', 'HELP_FGM', 'HELP_FGA', 'HELP_FG_PERC', 'MATCHUP_FTM', 'MATCHUP_FTA', 'SFL', 'MATCHUP_TIME_SEC'], 'rowSet': [['22019', 1629060, 'Rui Hachimura', 202329, 'Al-Farouq Aminu', 1, '6:45', 31.9, 4, 47, 1, 0, 0, 2, 5, 0.4, 0, 0, 0.0, 0, 0, 0, 0, 0, 0, 0, 404.7], ['22019', 1626168, 'Trey Lyles', 202329, 'Al-Farouq Aminu', 1, '7:31', 30.5, 3, 22, 0, 1, 0, 1, 3, 0.333, 1, 2, 0.5, 0, 0, 0, 0, 0, 0, 0, 451.1], ['22019', 1628384, 'OG Anunoby', 202329,

#### We can see that the key 'parameters' has a dictionary as its value. Let's loop over the keys of this dictionary and output the values.

This is practice for us to get our syntax correct for our nesting.

In [14]:
# Note the next level down of syntax
print('key-value pairs in a loop')
for k,v in nested[0]['parameters'].items():
    print("key:", k)
    print("value:",v,"\n")

key-value pairs in a loop
key: LeagueID
value: 00 

key: Season
value: 2019-20 

key: SeasonType
value: Regular Season 

key: PORound
value: 0 

key: PerMode
value: Totals 

key: DefTeamID
value: None 

key: OffTeamID
value: None 

key: OffPlayerID
value: None 

key: DefPlayerID
value: 202329 



In [15]:
# Output a single value from a selected key within the dict
# Note that we are THREE levels down to address an individual key
print('single value of a key')
print(nested[0]['parameters']['Season'])

single value of a key
2019-20


#### Now let's look at the resultSets key, with a list as its value.

1. Output the list that is the value for this key.

2. Loop over the list, to see the individual elements of it.

In [16]:
# Item #1
# output the value of the resultSets key
# again, note the levels to get down to it
print(nested[0]['resultSets'])

[{'name': 'SeasonMatchups', 'headers': ['SEASON_ID', 'OFF_PLAYER_ID', 'OFF_PLAYER_NAME', 'DEF_PLAYER_ID', 'DEF_PLAYER_NAME', 'GP', 'MATCHUP_MIN', 'PARTIAL_POSS', 'PLAYER_PTS', 'TEAM_PTS', 'MATCHUP_AST', 'MATCHUP_TOV', 'MATCHUP_BLK', 'MATCHUP_FGM', 'MATCHUP_FGA', 'MATCHUP_FG_PCT', 'MATCHUP_FG3M', 'MATCHUP_FG3A', 'MATCHUP_FG3_PCT', 'HELP_BLK', 'HELP_FGM', 'HELP_FGA', 'HELP_FG_PERC', 'MATCHUP_FTM', 'MATCHUP_FTA', 'SFL', 'MATCHUP_TIME_SEC'], 'rowSet': [['22019', 1629060, 'Rui Hachimura', 202329, 'Al-Farouq Aminu', 1, '6:45', 31.9, 4, 47, 1, 0, 0, 2, 5, 0.4, 0, 0, 0.0, 0, 0, 0, 0, 0, 0, 0, 404.7], ['22019', 1626168, 'Trey Lyles', 202329, 'Al-Farouq Aminu', 1, '7:31', 30.5, 3, 22, 0, 1, 0, 1, 3, 0.333, 1, 2, 0.5, 0, 0, 0, 0, 0, 0, 0, 451.1], ['22019', 1628384, 'OG Anunoby', 202329, 'Al-Farouq Aminu', 3, '5:56', 29.3, 5, 37, 1, 1, 1, 2, 4, 0.5, 1, 2, 0.5, 0, 0, 0, 0, 0, 0, 0, 355.6], ['22019', 201567, 'Kevin Love', 202329, 'Al-Farouq Aminu', 2, '5:17', 25.0, 4, 14, 0, 1, 0, 2, 3, 0.667, 0, 1,

In [17]:
# Second way of displaying item #1
nested[0]['resultSets']

[{'name': 'SeasonMatchups',
  'headers': ['SEASON_ID',
   'OFF_PLAYER_ID',
   'OFF_PLAYER_NAME',
   'DEF_PLAYER_ID',
   'DEF_PLAYER_NAME',
   'GP',
   'MATCHUP_MIN',
   'PARTIAL_POSS',
   'PLAYER_PTS',
   'TEAM_PTS',
   'MATCHUP_AST',
   'MATCHUP_TOV',
   'MATCHUP_BLK',
   'MATCHUP_FGM',
   'MATCHUP_FGA',
   'MATCHUP_FG_PCT',
   'MATCHUP_FG3M',
   'MATCHUP_FG3A',
   'MATCHUP_FG3_PCT',
   'HELP_BLK',
   'HELP_FGM',
   'HELP_FGA',
   'HELP_FG_PERC',
   'MATCHUP_FTM',
   'MATCHUP_FTA',
   'SFL',
   'MATCHUP_TIME_SEC'],
  'rowSet': [['22019',
    1629060,
    'Rui Hachimura',
    202329,
    'Al-Farouq Aminu',
    1,
    '6:45',
    31.9,
    4,
    47,
    1,
    0,
    0,
    2,
    5,
    0.4,
    0,
    0,
    0.0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    404.7],
   ['22019',
    1626168,
    'Trey Lyles',
    202329,
    'Al-Farouq Aminu',
    1,
    '7:31',
    30.5,
    3,
    22,
    0,
    1,
    0,
    1,
    3,
    0.333,
    1,
    2,
    0.5,
    0,
    0,
    0,
 

In [19]:
# Item #2
# Loop over the list that makes up this value
for i in nested[0]['resultSets']:
    print(i)
    print('\n')
    print("next i")

{'name': 'SeasonMatchups', 'headers': ['SEASON_ID', 'OFF_PLAYER_ID', 'OFF_PLAYER_NAME', 'DEF_PLAYER_ID', 'DEF_PLAYER_NAME', 'GP', 'MATCHUP_MIN', 'PARTIAL_POSS', 'PLAYER_PTS', 'TEAM_PTS', 'MATCHUP_AST', 'MATCHUP_TOV', 'MATCHUP_BLK', 'MATCHUP_FGM', 'MATCHUP_FGA', 'MATCHUP_FG_PCT', 'MATCHUP_FG3M', 'MATCHUP_FG3A', 'MATCHUP_FG3_PCT', 'HELP_BLK', 'HELP_FGM', 'HELP_FGA', 'HELP_FG_PERC', 'MATCHUP_FTM', 'MATCHUP_FTA', 'SFL', 'MATCHUP_TIME_SEC'], 'rowSet': [['22019', 1629060, 'Rui Hachimura', 202329, 'Al-Farouq Aminu', 1, '6:45', 31.9, 4, 47, 1, 0, 0, 2, 5, 0.4, 0, 0, 0.0, 0, 0, 0, 0, 0, 0, 0, 404.7], ['22019', 1626168, 'Trey Lyles', 202329, 'Al-Farouq Aminu', 1, '7:31', 30.5, 3, 22, 0, 1, 0, 1, 3, 0.333, 1, 2, 0.5, 0, 0, 0, 0, 0, 0, 0, 451.1], ['22019', 1628384, 'OG Anunoby', 202329, 'Al-Farouq Aminu', 3, '5:56', 29.3, 5, 37, 1, 1, 1, 2, 4, 0.5, 1, 2, 0.5, 0, 0, 0, 0, 0, 0, 0, 355.6], ['22019', 201567, 'Kevin Love', 202329, 'Al-Farouq Aminu', 2, '5:17', 25.0, 4, 14, 0, 1, 0, 2, 3, 0.667, 0, 1, 

#### This is very interesting, in that we can see that, although this is a list, there is only a single element in the list.

#### That element is a dictionary, as evidenced by the squiggly brackets framing the data.

While we went through a looping syntax to display this dict, let's address it directly.

Since this dict is the first element of the list (that makes up the value in the key-value pair), note the syntax to get down to it. This syntax builds on what we had above.

In [21]:
# address the list directly, notice the square bracket and index syntax,
# which is standard for addressing the elements of a list.
# this is the first item in the list
print(nested[0]['resultSets'][0])

{'name': 'SeasonMatchups', 'headers': ['SEASON_ID', 'OFF_PLAYER_ID', 'OFF_PLAYER_NAME', 'DEF_PLAYER_ID', 'DEF_PLAYER_NAME', 'GP', 'MATCHUP_MIN', 'PARTIAL_POSS', 'PLAYER_PTS', 'TEAM_PTS', 'MATCHUP_AST', 'MATCHUP_TOV', 'MATCHUP_BLK', 'MATCHUP_FGM', 'MATCHUP_FGA', 'MATCHUP_FG_PCT', 'MATCHUP_FG3M', 'MATCHUP_FG3A', 'MATCHUP_FG3_PCT', 'HELP_BLK', 'HELP_FGM', 'HELP_FGA', 'HELP_FG_PERC', 'MATCHUP_FTM', 'MATCHUP_FTA', 'SFL', 'MATCHUP_TIME_SEC'], 'rowSet': [['22019', 1629060, 'Rui Hachimura', 202329, 'Al-Farouq Aminu', 1, '6:45', 31.9, 4, 47, 1, 0, 0, 2, 5, 0.4, 0, 0, 0.0, 0, 0, 0, 0, 0, 0, 0, 404.7], ['22019', 1626168, 'Trey Lyles', 202329, 'Al-Farouq Aminu', 1, '7:31', 30.5, 3, 22, 0, 1, 0, 1, 3, 0.333, 1, 2, 0.5, 0, 0, 0, 0, 0, 0, 0, 451.1], ['22019', 1628384, 'OG Anunoby', 202329, 'Al-Farouq Aminu', 3, '5:56', 29.3, 5, 37, 1, 1, 1, 2, 4, 0.5, 1, 2, 0.5, 0, 0, 0, 0, 0, 0, 0, 355.6], ['22019', 201567, 'Kevin Love', 202329, 'Al-Farouq Aminu', 2, '5:17', 25.0, 4, 14, 0, 1, 0, 2, 3, 0.667, 0, 1, 

#### So what are the keys that make up this dictionary? Use the same syntax as above to list them.


In [22]:
# Note the next level down of syntax
# remember that we are still only looking at the first element of the list
print('key-value pairs in a loop')
for k,v in nested[0]['resultSets'][0].items():
    print("key:", k)
    print("value:",v,"\n")

key-value pairs in a loop
key: name
value: SeasonMatchups 

key: headers
value: ['SEASON_ID', 'OFF_PLAYER_ID', 'OFF_PLAYER_NAME', 'DEF_PLAYER_ID', 'DEF_PLAYER_NAME', 'GP', 'MATCHUP_MIN', 'PARTIAL_POSS', 'PLAYER_PTS', 'TEAM_PTS', 'MATCHUP_AST', 'MATCHUP_TOV', 'MATCHUP_BLK', 'MATCHUP_FGM', 'MATCHUP_FGA', 'MATCHUP_FG_PCT', 'MATCHUP_FG3M', 'MATCHUP_FG3A', 'MATCHUP_FG3_PCT', 'HELP_BLK', 'HELP_FGM', 'HELP_FGA', 'HELP_FG_PERC', 'MATCHUP_FTM', 'MATCHUP_FTA', 'SFL', 'MATCHUP_TIME_SEC'] 

key: rowSet
value: [['22019', 1629060, 'Rui Hachimura', 202329, 'Al-Farouq Aminu', 1, '6:45', 31.9, 4, 47, 1, 0, 0, 2, 5, 0.4, 0, 0, 0.0, 0, 0, 0, 0, 0, 0, 0, 404.7], ['22019', 1626168, 'Trey Lyles', 202329, 'Al-Farouq Aminu', 1, '7:31', 30.5, 3, 22, 0, 1, 0, 1, 3, 0.333, 1, 2, 0.5, 0, 0, 0, 0, 0, 0, 0, 451.1], ['22019', 1628384, 'OG Anunoby', 202329, 'Al-Farouq Aminu', 3, '5:56', 29.3, 5, 37, 1, 1, 1, 2, 4, 0.5, 1, 2, 0.5, 0, 0, 0, 0, 0, 0, 0, 355.6], ['22019', 201567, 'Kevin Love', 202329, 'Al-Farouq Aminu', 

#### These key-value pairs are getting down to the data that we would be seeking for our analysis.

What is in the value field for the "headers" key (the list) would be statistics names that we are analyzing. Were we to put this data into a row and column format (such as Excel or pandas), we can see that these are the column headers.

Let's directly address this list.

In [23]:
nested[0]['resultSets'][0]['headers'][0:5]

['SEASON_ID',
 'OFF_PLAYER_ID',
 'OFF_PLAYER_NAME',
 'DEF_PLAYER_ID',
 'DEF_PLAYER_NAME']

#### Let's look at the "rowset" key, and its values.

The square brackets at the very start and end of the value tell us that this is a list. Additionally, we can see that each of the individual elements is also a list, so this is a list of lists.

How do we address the top level list? Same as we just did for the "headers" key-value pair. Note that the 'S' in 'rowSet' is capitalized, so we don't run into an error that could throw us off.

In [26]:
print(nested[0]['resultSets'][0]['rowSet'])
print(nested[0]['resultSets'][0]['rowSet'])

[['22019', 1629060, 'Rui Hachimura', 202329, 'Al-Farouq Aminu', 1, '6:45', 31.9, 4, 47, 1, 0, 0, 2, 5, 0.4, 0, 0, 0.0, 0, 0, 0, 0, 0, 0, 0, 404.7], ['22019', 1626168, 'Trey Lyles', 202329, 'Al-Farouq Aminu', 1, '7:31', 30.5, 3, 22, 0, 1, 0, 1, 3, 0.333, 1, 2, 0.5, 0, 0, 0, 0, 0, 0, 0, 451.1], ['22019', 1628384, 'OG Anunoby', 202329, 'Al-Farouq Aminu', 3, '5:56', 29.3, 5, 37, 1, 1, 1, 2, 4, 0.5, 1, 2, 0.5, 0, 0, 0, 0, 0, 0, 0, 355.6], ['22019', 201567, 'Kevin Love', 202329, 'Al-Farouq Aminu', 2, '5:17', 25.0, 4, 14, 0, 1, 0, 2, 3, 0.667, 0, 1, 0.0, 0, 0, 0, 0, 0, 0, 0, 316.8]]
[['22019', 1629060, 'Rui Hachimura', 202329, 'Al-Farouq Aminu', 1, '6:45', 31.9, 4, 47, 1, 0, 0, 2, 5, 0.4, 0, 0, 0.0, 0, 0, 0, 0, 0, 0, 0, 404.7], ['22019', 1626168, 'Trey Lyles', 202329, 'Al-Farouq Aminu', 1, '7:31', 30.5, 3, 22, 0, 1, 0, 1, 3, 0.333, 1, 2, 0.5, 0, 0, 0, 0, 0, 0, 0, 451.1], ['22019', 1628384, 'OG Anunoby', 202329, 'Al-Farouq Aminu', 3, '5:56', 29.3, 5, 37, 1, 1, 1, 2, 4, 0.5, 1, 2, 0.5, 0, 0, 0,

#### OK, so this gives us the list of lists, now how do we address each of the list elements?

Just address them using the standard index notation.

For example, the first item in any list is just list[0]. So the first item in the rowset list is:

In [25]:
print(nested[0]['resultSets'][0]['rowSet'][0][0:6])

['22019', 1629060, 'Rui Hachimura', 202329, 'Al-Farouq Aminu', 1]


#### Finally, we might want to look at the individual data points in this list, which is the lowest level in our nested data structure.



In [27]:
# we can do this in a loop
for data_element in nested[0]['resultSets'][0]['rowSet'][0]:
    print(data_element)

# or we can address each data element directly
print("\ndirect address of data elements")
print(nested[0]['resultSets'][0]['rowSet'][0][0])
print(nested[0]['resultSets'][0]['rowSet'][0][1])

22019
1629060
Rui Hachimura
202329
Al-Farouq Aminu
1
6:45
31.9
4
47
1
0
0
2
5
0.4
0
0
0.0
0
0
0
0
0
0
0
404.7

direct address of data elements
22019
1629060


### To review what we have done here, let's look at the syntax of the various levels of the 'nested' data structure.

In [28]:
# At the top level, nested is a list of dicts.
for i in nested:  #loop over the elements of the list
    print(type(i))

<class 'dict'>
<class 'dict'>
<class 'dict'>
<class 'dict'>
<class 'dict'>
<class 'dict'>
<class 'dict'>


In [29]:
# to address each dict directly, we must use list index notation
# not using print() here, to cut down on the output
print(nested[0])
# display(nested[1])  # etc etc etc

{'resource': 'leagueseasonmatchups', 'parameters': {'LeagueID': '00', 'Season': '2019-20', 'SeasonType': 'Regular Season', 'PORound': 0, 'PerMode': 'Totals', 'DefTeamID': None, 'OffTeamID': None, 'OffPlayerID': None, 'DefPlayerID': 202329}, 'resultSets': [{'name': 'SeasonMatchups', 'headers': ['SEASON_ID', 'OFF_PLAYER_ID', 'OFF_PLAYER_NAME', 'DEF_PLAYER_ID', 'DEF_PLAYER_NAME', 'GP', 'MATCHUP_MIN', 'PARTIAL_POSS', 'PLAYER_PTS', 'TEAM_PTS', 'MATCHUP_AST', 'MATCHUP_TOV', 'MATCHUP_BLK', 'MATCHUP_FGM', 'MATCHUP_FGA', 'MATCHUP_FG_PCT', 'MATCHUP_FG3M', 'MATCHUP_FG3A', 'MATCHUP_FG3_PCT', 'HELP_BLK', 'HELP_FGM', 'HELP_FGA', 'HELP_FG_PERC', 'MATCHUP_FTM', 'MATCHUP_FTA', 'SFL', 'MATCHUP_TIME_SEC'], 'rowSet': [['22019', 1629060, 'Rui Hachimura', 202329, 'Al-Farouq Aminu', 1, '6:45', 31.9, 4, 47, 1, 0, 0, 2, 5, 0.4, 0, 0, 0.0, 0, 0, 0, 0, 0, 0, 0, 404.7], ['22019', 1626168, 'Trey Lyles', 202329, 'Al-Farouq Aminu', 1, '7:31', 30.5, 3, 22, 0, 1, 0, 1, 3, 0.333, 1, 2, 0.5, 0, 0, 0, 0, 0, 0, 0, 451.1],

In [30]:
# each of these elements themselves are dicts, with 3 keys
display(nested[0].keys())
display(nested[1].keys())

dict_keys(['resource', 'parameters', 'resultSets'])

dict_keys(['resource', 'parameters', 'resultSets'])

In [31]:
# each key-value pair is its own data structure
# loop over the dictionary, display the key and value pair
# commenting out the print() statements to cut down on the output
for k,v in nested[0].items():
    print("key:", k)
    print("value:",v,"\n")
    continue

key: resource
value: leagueseasonmatchups 

key: parameters
value: {'LeagueID': '00', 'Season': '2019-20', 'SeasonType': 'Regular Season', 'PORound': 0, 'PerMode': 'Totals', 'DefTeamID': None, 'OffTeamID': None, 'OffPlayerID': None, 'DefPlayerID': 202329} 

key: resultSets
value: [{'name': 'SeasonMatchups', 'headers': ['SEASON_ID', 'OFF_PLAYER_ID', 'OFF_PLAYER_NAME', 'DEF_PLAYER_ID', 'DEF_PLAYER_NAME', 'GP', 'MATCHUP_MIN', 'PARTIAL_POSS', 'PLAYER_PTS', 'TEAM_PTS', 'MATCHUP_AST', 'MATCHUP_TOV', 'MATCHUP_BLK', 'MATCHUP_FGM', 'MATCHUP_FGA', 'MATCHUP_FG_PCT', 'MATCHUP_FG3M', 'MATCHUP_FG3A', 'MATCHUP_FG3_PCT', 'HELP_BLK', 'HELP_FGM', 'HELP_FGA', 'HELP_FG_PERC', 'MATCHUP_FTM', 'MATCHUP_FTA', 'SFL', 'MATCHUP_TIME_SEC'], 'rowSet': [['22019', 1629060, 'Rui Hachimura', 202329, 'Al-Farouq Aminu', 1, '6:45', 31.9, 4, 47, 1, 0, 0, 2, 5, 0.4, 0, 0, 0.0, 0, 0, 0, 0, 0, 0, 0, 404.7], ['22019', 1626168, 'Trey Lyles', 202329, 'Al-Farouq Aminu', 1, '7:31', 30.5, 3, 22, 0, 1, 0, 1, 3, 0.333, 1, 2, 0.5, 0,

In [32]:
# Now we can address the 3 keys directly
# Notice the hierarchy of how we are addressing the dict, which is an element of the list
print(nested[0]['resource'])
print("\n")
print(nested[0]['parameters'])
print("\n")
print(nested[0]['resultSets'])

leagueseasonmatchups


{'LeagueID': '00', 'Season': '2019-20', 'SeasonType': 'Regular Season', 'PORound': 0, 'PerMode': 'Totals', 'DefTeamID': None, 'OffTeamID': None, 'OffPlayerID': None, 'DefPlayerID': 202329}


[{'name': 'SeasonMatchups', 'headers': ['SEASON_ID', 'OFF_PLAYER_ID', 'OFF_PLAYER_NAME', 'DEF_PLAYER_ID', 'DEF_PLAYER_NAME', 'GP', 'MATCHUP_MIN', 'PARTIAL_POSS', 'PLAYER_PTS', 'TEAM_PTS', 'MATCHUP_AST', 'MATCHUP_TOV', 'MATCHUP_BLK', 'MATCHUP_FGM', 'MATCHUP_FGA', 'MATCHUP_FG_PCT', 'MATCHUP_FG3M', 'MATCHUP_FG3A', 'MATCHUP_FG3_PCT', 'HELP_BLK', 'HELP_FGM', 'HELP_FGA', 'HELP_FG_PERC', 'MATCHUP_FTM', 'MATCHUP_FTA', 'SFL', 'MATCHUP_TIME_SEC'], 'rowSet': [['22019', 1629060, 'Rui Hachimura', 202329, 'Al-Farouq Aminu', 1, '6:45', 31.9, 4, 47, 1, 0, 0, 2, 5, 0.4, 0, 0, 0.0, 0, 0, 0, 0, 0, 0, 0, 404.7], ['22019', 1626168, 'Trey Lyles', 202329, 'Al-Farouq Aminu', 1, '7:31', 30.5, 3, 22, 0, 1, 0, 1, 3, 0.333, 1, 2, 0.5, 0, 0, 0, 0, 0, 0, 0, 451.1], ['22019', 1628384, 'OG Anunoby', 202329,

In [33]:
# We see that the 'resultSets' key has the values we need
# so output the value of the resultSets key
# again, note the levels to get down to it
print(nested[0]['resultSets'])

[{'name': 'SeasonMatchups', 'headers': ['SEASON_ID', 'OFF_PLAYER_ID', 'OFF_PLAYER_NAME', 'DEF_PLAYER_ID', 'DEF_PLAYER_NAME', 'GP', 'MATCHUP_MIN', 'PARTIAL_POSS', 'PLAYER_PTS', 'TEAM_PTS', 'MATCHUP_AST', 'MATCHUP_TOV', 'MATCHUP_BLK', 'MATCHUP_FGM', 'MATCHUP_FGA', 'MATCHUP_FG_PCT', 'MATCHUP_FG3M', 'MATCHUP_FG3A', 'MATCHUP_FG3_PCT', 'HELP_BLK', 'HELP_FGM', 'HELP_FGA', 'HELP_FG_PERC', 'MATCHUP_FTM', 'MATCHUP_FTA', 'SFL', 'MATCHUP_TIME_SEC'], 'rowSet': [['22019', 1629060, 'Rui Hachimura', 202329, 'Al-Farouq Aminu', 1, '6:45', 31.9, 4, 47, 1, 0, 0, 2, 5, 0.4, 0, 0, 0.0, 0, 0, 0, 0, 0, 0, 0, 404.7], ['22019', 1626168, 'Trey Lyles', 202329, 'Al-Farouq Aminu', 1, '7:31', 30.5, 3, 22, 0, 1, 0, 1, 3, 0.333, 1, 2, 0.5, 0, 0, 0, 0, 0, 0, 0, 451.1], ['22019', 1628384, 'OG Anunoby', 202329, 'Al-Farouq Aminu', 3, '5:56', 29.3, 5, 37, 1, 1, 1, 2, 4, 0.5, 1, 2, 0.5, 0, 0, 0, 0, 0, 0, 0, 355.6], ['22019', 201567, 'Kevin Love', 202329, 'Al-Farouq Aminu', 2, '5:17', 25.0, 4, 14, 0, 1, 0, 2, 3, 0.667, 0, 1,

In [34]:
# 'resultSets' has 3 keys, and we are interested in two of them
# the 'headers' key contains the column headers that we want for our data analysis
nested[0]['resultSets'][0]['headers']

['SEASON_ID',
 'OFF_PLAYER_ID',
 'OFF_PLAYER_NAME',
 'DEF_PLAYER_ID',
 'DEF_PLAYER_NAME',
 'GP',
 'MATCHUP_MIN',
 'PARTIAL_POSS',
 'PLAYER_PTS',
 'TEAM_PTS',
 'MATCHUP_AST',
 'MATCHUP_TOV',
 'MATCHUP_BLK',
 'MATCHUP_FGM',
 'MATCHUP_FGA',
 'MATCHUP_FG_PCT',
 'MATCHUP_FG3M',
 'MATCHUP_FG3A',
 'MATCHUP_FG3_PCT',
 'HELP_BLK',
 'HELP_FGM',
 'HELP_FGA',
 'HELP_FG_PERC',
 'MATCHUP_FTM',
 'MATCHUP_FTA',
 'SFL',
 'MATCHUP_TIME_SEC']

In [35]:
# the 'rowsets' key has the player full set of player data
print(nested[0]['resultSets'][0]['rowSet'])

[['22019', 1629060, 'Rui Hachimura', 202329, 'Al-Farouq Aminu', 1, '6:45', 31.9, 4, 47, 1, 0, 0, 2, 5, 0.4, 0, 0, 0.0, 0, 0, 0, 0, 0, 0, 0, 404.7], ['22019', 1626168, 'Trey Lyles', 202329, 'Al-Farouq Aminu', 1, '7:31', 30.5, 3, 22, 0, 1, 0, 1, 3, 0.333, 1, 2, 0.5, 0, 0, 0, 0, 0, 0, 0, 451.1], ['22019', 1628384, 'OG Anunoby', 202329, 'Al-Farouq Aminu', 3, '5:56', 29.3, 5, 37, 1, 1, 1, 2, 4, 0.5, 1, 2, 0.5, 0, 0, 0, 0, 0, 0, 0, 355.6], ['22019', 201567, 'Kevin Love', 202329, 'Al-Farouq Aminu', 2, '5:17', 25.0, 4, 14, 0, 1, 0, 2, 3, 0.667, 0, 1, 0.0, 0, 0, 0, 0, 0, 0, 0, 316.8]]


In [36]:
# the 'rowsets' key has the player full set of player data
print(nested[0]['resultSets'][0]['rowSet'])

[['22019', 1629060, 'Rui Hachimura', 202329, 'Al-Farouq Aminu', 1, '6:45', 31.9, 4, 47, 1, 0, 0, 2, 5, 0.4, 0, 0, 0.0, 0, 0, 0, 0, 0, 0, 0, 404.7], ['22019', 1626168, 'Trey Lyles', 202329, 'Al-Farouq Aminu', 1, '7:31', 30.5, 3, 22, 0, 1, 0, 1, 3, 0.333, 1, 2, 0.5, 0, 0, 0, 0, 0, 0, 0, 451.1], ['22019', 1628384, 'OG Anunoby', 202329, 'Al-Farouq Aminu', 3, '5:56', 29.3, 5, 37, 1, 1, 1, 2, 4, 0.5, 1, 2, 0.5, 0, 0, 0, 0, 0, 0, 0, 355.6], ['22019', 201567, 'Kevin Love', 202329, 'Al-Farouq Aminu', 2, '5:17', 25.0, 4, 14, 0, 1, 0, 2, 3, 0.667, 0, 1, 0.0, 0, 0, 0, 0, 0, 0, 0, 316.8]]


In [37]:
# and the value of the 'rowset' key is a list of lists, with each
# of the component lists containing a row of statistical data
# so we use the list index notation to address each element list
print(nested[0]['resultSets'][0]['rowSet'][0])

['22019', 1629060, 'Rui Hachimura', 202329, 'Al-Farouq Aminu', 1, '6:45', 31.9, 4, 47, 1, 0, 0, 2, 5, 0.4, 0, 0, 0.0, 0, 0, 0, 0, 0, 0, 0, 404.7]


In [38]:
# At the lowest level, each of these individual data elements makes up
# the list of statistics for the player matchup.
# So we again use the list index notation to address each individual statistic.
print(nested[0]['resultSets'][0]['rowSet'][0][0])

22019


#### What are your questions on addressing this nested data set?